### DISCUSSION TITLE
How codelens work and how to set up?

### DISCUSSION THREAD

**Discussion by BitInByte** [neovim/neovim#24791] [Q&A]
Hello guys, hope you're having an amazing start of a weekend :D 

I've been trying to mess around with codelens and I've been searching through a lot of articles and even on the help documentation on how to set up properly. I know that there's a bunch of lsp's that are not supported yet but, to be honest, none of my lsp's are working. 

My question is, does codelens work by default? If not, how could we enable it in order to enable for the lsp's that support codelens? 

Thanks y'all for any help and support :D

**Comment by mmj2023** [comment]
You can look at lazyvim's (a neovim distribuition) documentation for setting up their lsp's there is an example config there however you will need to do some chages to it otherwise it might not work as expected.

**Comment by maxmx03** [comment]
## Codelens Setup

To enable Codelens in Neovim, you'll need to set up an autocmd that refreshes the Codelens when a language server attaches to a buffer.

```lua
autocmd('LspAttach', {
  callback = function(args)
    local bufnr = args.buf
    local client = vim.lsp.get_client_by_id(args.data.client_id)

    if client and client:supports_method 'textDocument/codeLens' then
      vim.lsp.codelens.refresh()
      vim.api.nvim_create_autocmd({ 'BufEnter', 'CursorHold', 'InsertLeave' }, {
        buffer = bufnr,
        callback = vim.lsp.codelens.refresh,
      })
    end
  end,
})
```

## Language Server Configuration

After setting up the autocmd, you must configure your specific language server to enable Codelens. For lua_ls, you'll need to add `codeLens.enable = true` to its settings. [Codelens Configuration](https://luals.github.io/wiki/settings/#codelensenable)

```lua
vim.lsp.config('lua_ls', {
   -- ...
   settings = {
    Lua = {
      codeLens = {
        enable = true,
      },
      -- ...
    },
  },
})
vim.lsp.enable('lua_ls')
```

## Example

The params contain an inlayhints, while `references: 5` indicates a Codelens display.

![codelens](https://github.com/user-attachments/assets/d4468672-3acd-4d6e-995a-c812110c4e98)
